src/libcore/any.rs

```diff
@@ -450,3 +450,22 @@ impl TypeId {
         }
     }
 }
+
+/// Returns the name of a type as a string.
+///
+/// # Note
+///
+/// This is intended for diagnostic use. The exact contents and format of the
+/// string are not specified, other than being a best-effort description of
+/// the type. For example, `type_name::<Option<String>>()` could return the
+/// string `"Option<String>"` or
+/// `"core::option::Option<alloc::string::String>"`, but not `"foobar"`.
+///
+/// The current implementation uses the same infrastructure as compiler
+/// diagnostics and debuginfo but this is not guaranteed.
+#[stable(feature = "type_name", since = "1.36.0")]
+pub fn type_name<T: ?Sized>() -> &'static str {
+    unsafe {
+        intrinsics::type_name::<T>()
```

Comment:
Please also make the `type_name` intrinsic safe. We have a list of safe intrinsics flying around somewhere (grep for `"size_of"`, including the quotes).

Reply:
Done!